### PR TITLE
store outputs in a content-addressed store

### PR DESCRIPTION
**Pull request by BrianHicks**
That lets us avoid doing work, like so:

```
$ roc run examples/hello/rbt.roc
🔨 Rebuilding host...
2022-08-18T22:34:02.255Z DEBUG [host::coordinator] preparing to run job d2a0eee0ece469cb
2022-08-18T22:34:02.259Z TRACE [host::store] moving out into collection path
2022-08-18T22:34:02.259Z DEBUG [host::store] moving f8ccb17e33d1f54fdc0761c864ef8d5881430bc55e887e0647ffc951b96d8032 into store

$ roc run examples/hello/rbt.roc
🔨 Rebuilding host...
2022-08-18T22:34:11.889Z DEBUG [host::coordinator] preparing to run job d2a0eee0ece469cb
2022-08-18T22:34:11.889Z DEBUG [host::coordinator] already had output of this job; skipping
```

I know this code will need to be changed for this to work in parallel, but we need this logic anyway so I think that's OK for the moment.

Based on #32; merge that first!

### CODE REVIEW

File: src/store.rs [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+        let final_hash = hasher.finalize().to_hex().to_string();
+        let final_location = self.root.join(&final_hash);
+
+        log::debug!("moving {} into store", &final_hash);
+
+        // important: at this point we need to take ownership of the tempdir so
+        // that it doesn't get automatically removed when it's dropped. We've
+        // so far avoided that to avoid leaving temporary directories laying
+        // around in case of errors.
+        std::fs::rename(temp.into_path(), &final_location)
```
Comment (bhansconnect) [inline src/store.rs:170]: Instead of having a temporary directory that is in `/tmp`, would it make more sense to put in the store with a prefix or suffix, that way if something goes wrong, you can still look at what was outputted and potentially debug better?

File: src/coordinator.rs [hunk 2]
```diff
@@ -25,18 +42,30 @@ impl Coordinator {
...
+            let workspace = Workspace::create(&self.workspace_root, job)
+                .with_context(|| format!("could not create workspace for job {}", job.id))?;
+
+            runner.run(job, &workspace).context("could not run job")?;
+
+            self.store
+                .store_from_workspace(job, workspace)
+                .context("could not store job output")?;
+        } else {
+            log::debug!("already had output of this job; skipping");
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/coordinator.rs:67]: actually, I'd expect this to be the more common case by far. If you run the build twice, we see that we have a matching output path for the build's input hash and can skip running.

File: src/job.rs [hunk 3]
```diff
@@ -13,9 +14,35 @@ impl From<u64> for Id {
...
 impl From<&glue::Job> for Id {
-    fn from(job: &glue::Job) -> Self {
+    /// We don't care about order in some places (e.g. output file) while we do
+    /// in others (e.g. command arguments.) The hash should reflect this!
+    ///
+    /// Note: this data structure is going to grow the ability to refer to other
+    /// jobs as soon as it's feasible. When that happens, a depth-first search
+    /// through the tree rooted at `top_job` will probably suffice.
+    fn from(top_job: &glue::Job) -> Self {
+        // TODO: is this the best hash for this kind of data? Should we find
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/job.rs:18]: yeahhhh, I noticed the admonition that the implementation can change anytime later. What do you reckon would be a good target? I was thinking xxhash but I'm not sure… it's certainly going to be a lot of short-ish input!

[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/job.rs:18]: issued at https://github.com/roc-lang/rbt/issues/38. Nobody's depending on these hashes yet, so just to keep things moving I'd like to defer that until the current PR stack is all closed out. Sound OK to you?

File: src/job.rs [hunk 4]
```diff
@@ -13,9 +14,35 @@ impl From<u64> for Id {
...
-        job.hash(&mut hasher);
+
+        let job = &top_job.f0;
+
+        // TODO: when we can get commands from other jobs, we need to hash the
+        // other tool and job instead of relying on the derived `Hash` trait
+        // for this.
+        job.command.hash(&mut hasher);
+
+        // TODO: input file hashes need to change this hash. We cannot do that
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/job.rs:29]: we need to at least know file metadata for inputs in advance (so we know when to re-run the build.) We also need to get the hashes (so we know when metadata changes don't indicate actual changes.) That's out of scope here, but happened in #37.

We don't need to do the same things for output files because they cannot be known in advance (but this PR takes care of storing them in a content-addressable way.)

File: src/store.rs [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+            // buffer. Gonna have to do this by hand at some point to take
+            // advantage of the algorithm's designed speed.
+            std::io::copy(&mut file, &mut hasher).with_context(|| {
+                format!("could not read `{}` to calculate hash", source.display())
+            })?;
+
+            ///////////////////////////////////////////////
+            // Step 3: make sure any parent paths exist  //
+            ///////////////////////////////////////////////
+            let mut ancestors: Vec<&Path> = source.ancestors().skip(1).collect();
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:107]: ancestors for `a/b/c" will look like `["a/b/c", "a/b", "a", ""]` ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=4c1ba455ad1061470830c74e1c5bd7c7)) From those:

- `skip(1)` because we don't want "a/b/c" because that's the full path—if we create a directory there, we'll conflict with the output.
- `pop()` because we also don't want the path's final base (in this case always "" since we've determined the path is relative already.)

File: src/store.rs [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+            ancestors.pop(); // we've made sure this is relative, so the first item is ""
+            ancestors.reverse(); // go `[a, a/b, a/b/c]` instead of `[a/b/c, a/b, a]`
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:109]: In that case we'd need to pop `""` from the front (see my other comment) and would need to use `remove()` with an `O(n)` runtime, or use a `VecDeque`. `reverse()` happens in place. The docs don't specify a runtime, but I'd imagine it's like half the length since we'd be doing n/2 swaps.

But in the bigger picture, I don't expect these will be more than a handful of directories deep. There's no need for extremely deep nesting since output directories do not clash with each other in the store. The `n` here would small enough to not matter, unless I've badly misunderstood something about how iterators operate!

File: src/store.rs [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                if output_dirs.contains(&ancestor) {
+                    continue;
+                }
+
+                log::trace!(
+                    "creating {} in {}",
+                    &ancestor.display(),
+                    temp.path().display()
+                );
+                std::fs::create_dir(temp.path().join(&ancestor)).with_context(|| {
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:123]: that might simplify the code some, yeah, but at a performance cost:

- we'd be doing a filesystem read for each item in the ancestor chain for every item in the outputs to check if we need to make the directory
- we'd have to walk all the output paths again later to make the directories read-only after all the files are situated within them

That approach would save us some memory and a little bit of complexity, but make disk performance much worse. I could still see it being *ok* though, since ideally we will be avoiding rebuilds (and the subsequent workspace consumptions) as much as possible. The exception would be on the first build, where we'll be rebuilding everything.

File: src/store.rs (hunk 1 above)
Comment (BrianHicks) [reply src/store.rs:170]: Yeah, that might make sense. Easy to implement, too!

File: src/coordinator.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/coordinator.rs:67]: Ah, I miss understood this. I was thinking within a single build, which is why it seemed weird. Over multiple builds makes sense.

File: src/job.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/job.rs:18]: Yeah, just fix later.

File: src/job.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/job.rs:29]: Ack.

File: src/store.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:107]: ahh, different sides of the iterator.

File: src/store.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:109]: Ah, so I guess you could do `source.ancestors().skip(1).rev().skip(1)` and avoid the collecting all together.

But that is also harder to understand, so I totally get it if you prefer this setup.

File: src/store.rs [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+impl Store {
+    pub fn new(root: PathBuf) -> Result<Self> {
+        let inputs_to_content = match File::open(&root.join("inputs_to_content.json")) {
+            Ok(file) => {
+                let reader = BufReader::new(file);
+                serde_json::from_reader(reader)
+                    .context("could not deserialize mapping from inputs to content")?
+            }
+            Err(err) if err.kind() == std::io::ErrorKind::NotFound => HashMap::default(),
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:32]: yeah, the first time we load the JSON file from disk it won't exist and we should just use an empty one and then persist it later.

This will go away later when we're using an actual DB (which I'd like to do ASAP, probably sled.)

File: src/store.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:109]: I'm not sure either of those is particularly intuitive. I should probably add a comment on this similar to the one I left above!

File: src/store.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:123]: > we'd be doing a filesystem read for each item in the ancestor chain

Not exactly. create_dir_all, starts at the child node and works backwards towards the ancestors as necessary. So to create `a/b/c/d/e` where `a/b` already exists, it will:
1. try and fail to create `a/b/c/d/e`
1. try and fail to create `a/b/c/d`
1. succeed at creating `a/b/c`
1. succeed at creating `a/b/c/d`
1. succeed at creating `a/b/c/d/e`

Not sure the cost of failing to create a dir because it's parent doesn't exist, but assuming that is cheap, this should be fast.

> we'd have to walk all the output paths again later to make the directories read-only after all the files are situated within them

True, the cost here is probably fs dependent. I would assume not marking anything read only and then later marking everything read only would have low cost. 1, most things should be in fs cache. 2, you would only be traversing metadata that should be stored pretty compactly. Also, I would assume the cost to hash and copy files would overshadow all of this anyway.


Just to clarify, in this scenario, either everything or nothing should exist, correct? Since either we ran before and have all the outputs, or we haven't and need to build all directories. If this is correct, your most efficient option would be to process all ancestors of all files at once and just do a de-duplication of the list of ancestors.

File: src/store.rs [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+            log::trace!("moving {} into collection path", &source.display());
+            let out = temp.path().join(&source);
+            std::fs::rename(workspace.join(&source), &out).with_context(|| {
+                format!(
+                    "could not move `{}` from workspace to store",
+                    source.display()
+                )
+            })?;
+
+            Self::make_readonly(&out).with_context(|| {
```
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:145]: Ack.

File: src/store.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:123]: > Not exactly. create_dir_all, starts at the child node and works backwards towards the ancestors as necessary.

Ah, that's on me for assuming that `create_dir_all` would start from the root. Thanks for clarifying.

> Also, I would assume the cost to hash and copy files would overshadow all of this anyway.

Hashing, sure, but we're renaming the files instead of copying them since the original bytes would be reaped anyway when the build workspace was deleted.

> Just to clarify, in this scenario, either everything or nothing should exist, correct?

Yeah, that's right.

> If this is correct, your most efficient option would be to process all ancestors of all files at once and just do a de-duplication of the list of ancestors.

Let me rephrase this to make sure I understand what you're suggesting:

1. calculate the hash based on the files
2. exit early if we already have the hash
3. do the filesystem work, deduplicating ancestors

I ask because we already are effectively deduplicating ancestors—we're leaning heavily on the fact that the output directory won't exist in advance, so we don't need to check if some parent directory exists in advance of creating it. I wonder about combining our suggestions, such that the "do the filesystem work" above looks like:

1. create the temporary collection directory
2. create all the directories
3. move all the files, marking them as read-only
4. mark all the directories as read-only
5. move the collection directory into the store

(I want to insist on steps 1 and 5 for safety, even though they'll add just a little overhead. It'd be really bad to have a corrupt path in the CAS!)

Anyway, I think this could end up much nicer, both from an efficiency perspective and an ease-of-understanding one. We could reuse the approach for create-move-mark here, or use some other kind of filesystem walking like `create_dir_all`. (I think the deciding factor maybe should be approachability instead of speed here, although with some reasonable documentation of assumptions I think the current approach could be fairly understandable.)

[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:123]: oh also, this maybe should be a separate refactor in a new PR! The current approach works OK—this would get that last little bit of performance out and make it easier to understand, but I don't think it makes sense to block the rest of the outstanding PRs on that work.

[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:123]: Yeah, I am fine with either solution, just feel there are ways to simplify the readability here. Anyway, not needed for this PR for sure.

File: src/store.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (bhansconnect) [reply src/store.rs:32]: Ack

**Review by bhansconnect — APPROVED** [review]

File: src/store.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:123]: yeah. I think I do want to change it. Might as well apply the guiding principles (understandable, then approachable, then fast) to the code as well as the system's behavior. Thank you for this excellent review!

File: src/cli.rs [hunk 10]
```diff
@@ -14,20 +15,23 @@ pub struct Cli {
...
-    isolator_root: PathBuf,
+    root_dir: PathBuf,
 }
 
 impl Cli {
     pub fn run(&self) -> Result<()> {
         let rbt = Self::load();
 
-        let mut coordinator = Coordinator::default();
+        let store = Store::new(self.root_dir.join("store")).context("could not open store")?;
```
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/cli.rs:25]: whoops, I missed this earlier. Yes, the way the directories get passed around in here is getting messy and it's gonna have to be cleaned up before all the info lives in a database.

[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/cli.rs:25]: gonna merge now, though, probably 😅

File: src/store.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (BrianHicks) [reply src/store.rs:109]: BTW, turns out we can't do this: `rev` is only implemented for double-ended iterators, which `ancestors` is not. ☹️
